src/AzureExtension/DataModel/DataObjects/Project.cs

```diff
@@ -1,186 +1,232 @@
+    {
+        return new Project
+        {
+            InternalId = project.Id.ToString(),
+            Name = project.Name ?? string.Empty,
+            Description = project.Description ?? string.Empty,
+            OrganizationId = organizationId,
+            TimeUpdated = DateTime.Now.ToDataStoreInteger(),
+        };
+    }
+
+    private static Project CreateFromTeamProject(TeamProjectReference project, long organizationId)
+    {
+        return new Project
+        {
+            InternalId = project.Id.ToString(),
+            Name = project.Name ?? string.Empty,
+            Description = project.Description ?? string.Empty,
+            OrganizationId = organizationId,
+            TimeUpdated = DateTime.Now.ToDataStoreInteger(),
+        };
+    }
+
+    public static Project AddOrUpdateProject(DataStore dataStore, Project project)
+    {
+        // Check for existing Project data.
+        var existingProject = GetByInternalId(dataStore, project.InternalId);
+        if (existingProject is not null)
+        {
+            // Many of the same Project records will be created on a sync, and to
+            // avoid unnecessary updating and database operations for data that
+            // is extremely unlikely to have changed in any significant way, we
+            // will only update every UpdateThreshold amount of time.
+            if ((project.TimeUpdated - existingProject.TimeUpdated) > _updateThreshold)
+            {
+                project.Id = existingProject.Id;
+                dataStore.Connection!.Update(project);
+                project.DataStore = dataStore;
+                return project;
+            }
+            else
+            {
+                return existingProject;
+            }
+        }
+
+        // No existing project, add it.
+        project.Id = dataStore.Connection!.Insert(project);
+        project.DataStore = dataStore;
+        return project;
+    }
+
+    // Direct Get always returns a non-null object for reference in other objects.
+    public static Project Get(DataStore? dataStore, long id)
+    {
+        if (dataStore == null)
+        {
+            return new Project();
+        }
+
+        var project = dataStore.Connection!.Get<Project>(id);
+        if (project != null)
+        {
+            project.DataStore = dataStore;
+        }
+
+        return project ?? new Project();
+    }
+
+    public static Project? Get(DataStore? dataStore, string projectName, string organizationName)
+    {
+        if (dataStore == null)
+        {
+            return null;
+        }
+
+        // This is not guaranteed to only return one result, such as the case of multiple
+        // organization records with different connections.
+        var sql = @"SELECT * FROM Project AS P WHERE (P.Name = @Name OR P.InternalId = @Name) AND P.OrganizationId IN (SELECT Id FROM Organization WHERE Organization.Name = @Org)";

```

Comment:
Please use parametrized SQL query inputs.  I understand this isn't the web, but we should protect against SQL injection.

Reply:
Every SQL query is already parameterized precisely to protect against injection. It's literally the next lines in the code. That string is not interpolated with values directly, they're set as parameters and passed to the SQL call as such.

Reply:
ic ic.  I didn't notice the next line.
Back in my first job I was conditioned to use [SqlCommand.Parameters](https://learn.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlcommand.parameters?view=netframework-4.8.1) to avoid SQL injection.

Knee jerk reaction.